executor/executor.c

```diff
@@ -0,0 +1,492 @@
+	// Wait for thread to exit and then remove everything associated with the thread
+    pthread_join(tid, NULL);
+    pthread_mutex_unlock(&mutex);
+    pthread_mutex_destroy(&mutex);
+    pthread_cond_destroy(&cond);
+}
+
+
+/**
+ *  Begins the given game mode and calls setup and main appropriately. Will run main forever.
+ * 
+ *  Behavior: This is a blocking function and will block the calling thread forever.
+ *  This should only be run as a separate thread.
+ * 
+ *  Inputs:
+ *      args: string of the mode to start running
+ */
+void* run_mode(void *args) {
+	//Set up the arguments to the threads that will run the setup and main threads
+    char *mode = (char *)args;
+    char setup_str[20], main_str[20];
+    sprintf(setup_str, "%s_setup", mode);
+    sprintf(main_str, "%s_main", mode);
+    thread_args_t setup_args = { setup_str, NULL, mode, &setup_time, 0 };
+    thread_args_t main_args = { main_str, NULL, mode, &main_interval, 1 };
+	
+	//Ensure that SIGINT is only handled by main thread
+    sigset_t set;
+    sigemptyset(&set);
+    sigaddset(&set, SIGINT);
+    pthread_sigmask(SIG_BLOCK, &set, NULL);
+	
+	//Run the setup function on a timeout, followed by main function on a loop
+    run_function_timeout(&setup_args);
+    run_py_function(&main_args);
+    return NULL;
+}
+
+
+/**
+ *  Stops the current mode and makes sure to cleanly exit any currenly running Python threads.
+ */
+void stop_mode() {
+    // Must acquire the GIL before calling C API functions
+    PyGILState_STATE gstate = PyGILState_Ensure();
+    log_runtime(DEBUG, "Got GIL");
+    
+    // Cancel the main Python thread by sending a TimeoutError
+    PyThreadState_SetAsyncExc((unsigned long) mode_thread, PyExc_TimeoutError);
+    PyErr_Clear();
+
+    // Get all threads started by student
+    PyObject* running_actions = PyObject_GetAttrString(pRobot, "running_actions");
+    if (running_actions == NULL) {
+        PyErr_Print();
+        log_runtime(ERROR, "getting running actions failed");
+        return;
+    }
+    PyObject* actions = PyMapping_Values(running_actions);
+    Py_DECREF(running_actions);
+    if (actions == NULL) {
+        PyErr_Print();
+        log_runtime(ERROR, "Getting running threads failed");
+        return;
+    }
+
+    // Iterate through all threads started
+    Py_ssize_t num_actions = PyList_Size(actions);
+    for (int i = 0; i < num_actions; i++) {
+        PyObject* thread = PyList_GetItem(actions, i);
+        if (thread == NULL) {
+            PyErr_Print();
+            log_runtime(ERROR, "Get item from running_actions failed");
+            continue;
+        }
+
+        // Ensure we only cancel threads that are still alive
+        PyObject* alive = PyObject_CallMethod(thread, "is_alive", NULL);
+        if (alive == NULL) {
+            PyErr_Print();
+            log_runtime(ERROR, "Couldn't get whether thread is alive");
+            continue;
+        }
+        else if (PyObject_Not(alive)) {
+            continue;

```

Comment:
Is a `Py_DECREF(alive);` necessary here?

Reply:
yea I forgot to decref stuff in that function. I'll fix it